### PR TITLE
Candidate for version 3.2.5

### PR DESCRIPTION
**Pull request by kushti**
Candidate for release 3.2.5. It contains:

* Support for offchain boxes spending ( #1032  )
* Downloading logic fix for the case when best full block is not on the best header-chain
* Improved storage rent tests ( #1114 )

Contributors: @jasondavies , @kushti, @polarker

### CODE REVIEW

File: src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/ToDownloadProcessor.scala [hunk 1]
```diff
@@ -37,20 +37,22 @@ trait ToDownloadProcessor extends BasicReaders with ScorexLogging {
...
-            val toDownload = requiredModifiersForHeader(bestHeaderAtThisHeight)
-              .filter(m => condition(m._2))
-            continuation(height + 1, acc ++ toDownload)
-          case None => acc
+        val headersAtThisHeight = headerIdsAtHeight(height).flatMap(id => typedModifierById[Header](id))
+
+        if (headersAtThisHeight.nonEmpty) {
+          val toDownload = headersAtThisHeight.flatMap(h => requiredModifiersForHeader(h))
+                       .filter(m => condition(m._2))
+                    continuation(height + 1, acc ++ toDownload)
```
Comment (jasondavies) [inline src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/ToDownloadProcessor.scala:53]: Needs indentation fix.

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala [hunk 2]
```diff
@@ -420,63 +427,68 @@ class ErgoWalletActor(settings: ErgoSettings, boxSelector: BoxSelector)
...
+    *
+    * @return generated transaction or an error
     */
   private def generateTransactionWithOutputs(requests: Seq[TransactionGenerationRequest],
                                              inputsRaw: Seq[String],
                                              dataInputsRaw: Seq[String]): Try[ErgoTransaction] = Try {
+
+    // A helper which converts Base16-encoded boxes to ErgoBox instances
+    def stringsToBoxes(strings: Seq[String]): Seq[ErgoBox] =
+      strings.map(in => Base16.decode(in).flatMap(ErgoBoxSerializer.parseBytesTry)).map(_.get)
```
Comment (oskin1) [inline src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala:447]: Use `Seq(Try()).collect { case Success(x) => x }` at least.

File: src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/ToDownloadProcessor.scala (hunk 1 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/history/storage/modifierprocessors/ToDownloadProcessor.scala]: fixed!

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala (hunk 2 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala:447]: @oskin1 generateTransactionWithOutputs body is wrapped in Try, so .get isn't a problem

File: ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala [hunk 3]
```diff
@@ -27,7 +27,7 @@ class ErgoInterpreter(params: ErgoLikeParameters)(implicit IR: IRContext)
     * Check that expired box is spent in a proper way
     *
     * @param box           - box being spent
-    * @param output        - newly created box
+    * @param output        - newly created box when storage fee covered, otherwise any output box
```
Comment (aslesarenko) [inline ergo-wallet/src/main/scala/org/ergoplatform/wallet/interpreter/ErgoInterpreter.scala:30]: This is not related to the changes, but I cannot understand the logic of this method. 
Looks like something wrong, or comments would be helpfull.

File: src/main/resources/reference.conf [hunk 4]
```diff
@@ -339,7 +339,7 @@ scorex {
     getPeersInterval = 2m
 
     # Number of network connections
-    maxConnections = 20
+    maxConnections = 30
```
Comment (aslesarenko) [inline src/main/resources/reference.conf:342]: Any comments on why the parameters where increased?

File: src/main/scala/org/ergoplatform/network/ErgoNodeViewSynchronizer.scala [hunk 5]
```diff
@@ -77,6 +77,26 @@ class ErgoNodeViewSynchronizer(networkControllerRef: ActorRef,
     }
   }
 
+  //todo: pull back to Scorex
+  override protected def checkDelivery: Receive = {
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/network/ErgoNodeViewSynchronizer.scala:81]: Please describe why this method require override, and why it should be pulled back to Scorex?

File: src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala [hunk 6]
```diff
@@ -314,12 +314,19 @@ class ErgoWalletActor(settings: ErgoSettings, boxSelector: BoxSelector)
...
+    * (and not spent offchain, but that is ensured by offChainRegistry).
+    * This filter is used when the wallet is going through its boxes to assemble a transaction.
     */
-  private val onChainFilter: FilterFn = (trackedBox: TrackedBox) => trackedBox.chainStatus.onChain &&
-    offChainRegistry.onChainBalances.exists(_.id == encodedBoxId(trackedBox.box.id))
+  private val walletFilter: FilterFn = (trackedBox: TrackedBox) => {
+    if(trackedBox.chainStatus.onChain) {
+      offChainRegistry.onChainBalances.exists(_.id == trackedBox.boxId)
+    } else {
+      true
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/nodeView/wallet/ErgoWalletActor.scala:327]: Please describe what this true mean?

**Pull request merged**